spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/MetricsEndpoint.java

```diff
@@ -112,15 +113,22 @@ private Tag parseTag(String tag) {
 		return Tag.of(parts[0], parts[1]);
 	}
 
-	private void collectMeters(List<Meter> meters, MeterRegistry registry, String name,
+	private List<Meter> findFirstMatchingMeters(MeterRegistry registry, String name,
 			Iterable<Tag> tags) {
 		if (registry instanceof CompositeMeterRegistry) {
-			((CompositeMeterRegistry) registry).getRegistries()
-					.forEach((member) -> collectMeters(meters, member, name, tags));
+			return ((CompositeMeterRegistry) registry).getRegistries().stream()
+					.map((r) -> findFirstMatchingMeters(r, name, tags))
+					.filter((match) -> !match.isEmpty()).findFirst()
+					.orElse(Collections.emptyList());
+
 		}
 		else {
-			meters.addAll(registry.find(name).tags(tags).meters());
+			Collection<Meter> metersFound = registry.find(name).tags(tags).meters();
+			if (!metersFound.isEmpty()) {
+				return new ArrayList<>(metersFound);
```

Comment:
No need to wrap this. The list returned is already an immutable copy.

Reply:
Thanks @jkschneider , the reason I am wrapping it into `List` is to avoid returning `Collection` from this method. 

`Collection<Meter> metersFound = registry.find(name).tags(tags).meters();`

If we return `Collection`, the `MetricEndpoint` methods should be refactored so that `Collection` could be pass around instead of `List` interface **OR** the return type should be downcasted to `List` (which is dangerous as we dont have control on `MeterRegistry#find`)

Changing `MetricEndpoint` methods to pass `Collection` is minor refactoring, with following highlight.

https://github.com/spring-projects/spring-boot/blob/13fef7d6cdaf5242c333f927dc0c2713823731ba/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/MetricsEndpoint.java#L93 

will be change to 

`meters.iterator().next().getId();`

Please confirm should I go ahead and do this minor change?

Reply:
The minor refactoring seems preferable to me to generating extra garbage in this case, as minor as that will be.

Reply:
Thanks @jkschneider, pls review.